[GOBBLIN-1398] load system specific hive config to support multiple h…

jhsenjaliya opened the pull request:
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!


### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1398


### Description
- [x] Here are some details about my PR, including screenshots (if applicable):

GOBBLIN-1308 takes care of ability to connect to remote and secure hive metastore, but it still requires the management of hive-site to be provided to be container local file. 

When it comes to multiple hive clusters this manual management approach would not work, and requires special feature of providing system specific hive-site.xml without namespace collision.

This ticket aims to do following things
1) define a way to provide remote system configuration ( keep giving flat config is more cumbersome )
2) based on system config and feature flag, copy config files to container local path automatically.
3) when creating metastoreClient, pick up the right config for the requested system ( identified by the metastore URI )

 

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason: Needs to add unit test once reviewed.


### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

jhsenjaliya committed:
[GOBBLIN-1398] load system specific hive config to support multiple hive metastores

autumnust commented:
Hi Jay, thanks for this patch. Starting to look at it. 
Can you also update the your PR description for more context of what you changed as well as replace all Gobblin-XXX with the actual ticket number ? Thanks.

autumnust commented on gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinYarnAppLauncher.java:
```diff
@@ -731,6 +752,25 @@ private void addContainerLocalResources(ApplicationId applicationId) throws IOEx
+
+    // handles all system configs files automatically, so it does not has to be mentioned explicitly as values of CONTAINER_FILES_LOCAL_KEY
+    if (this.config.hasPath(GobblinYarnConfigurationKeys.CONTAINER_COPY_SYSTEM_CONFIGS) && this.config.getBoolean(GobblinYarnConfigurationKeys.CONTAINER_COPY_SYSTEM_CONFIGS)) {
+      Path appFilesDestDir = new Path(containerWorkDir, GobblinYarnConfigurationKeys.APP_FILES_DIR_NAME);
```
It seems it needs to be named differently since it is under `containerWorkDir`

autumnust commented on gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinYarnAppLauncher.java:
```diff
@@ -731,6 +752,25 @@ private void addContainerLocalResources(ApplicationId applicationId) throws IOEx
+            addAppLocalFiles(e.getValue(), e.getKey()+"-", appFilesDestDir,
+                    Optional.<Map<String, LocalResource>>absent(), localFs);
+          } catch (IOException ioException) {
+            ioException.printStackTrace();
```
Same logging ask as before

autumnust commented on gobblin-utility/src/main/java/org/apache/gobblin/util/ConfigUtils.java:
```diff
@@ -569,4 +608,124 @@ public static Config resolveEncrypted(Config config, Optional<String> encConfigP
+
+    for (String systemName : config.getObject(GOBBLIN_SYNC_SYSTEMS_KEY).keySet()){
+      Config sysConfig = config.getConfig(GOBBLIN_SYNC_SYSTEMS_KEY).getConfig(systemName);
+      if(ConfigUtils.getBoolean(sysConfig, IS_SECURE_KEY, false) && ConfigUtils.getBoolean(sysConfig, IS_TOKEN_MGMT_ENABLED_KEY, false) && sysConfig.hasPath(HIVE_SYSTEM_KEY)){
```
Not sure it is properly to leak contents of config in a utility class

autumnust reviewed (changes requested):
As a general comment, is that possible to add some unit tests for your changes?

jhsenjaliya commented on gobblin-utility/src/main/java/org/apache/gobblin/util/ConfigUtils.java:
```diff
@@ -34,6 +34,9 @@
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
+
+import com.google.common.collect.*;
```
ya, need to setup my new intelliJ configs, will update.

jhsenjaliya commented on gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinYarnAppLauncher.java:
```diff
@@ -771,20 +811,33 @@ private void addAppJars(String jarFilePathList, Optional<Map<String, LocalResour
+    }
+  }
+
+  private void addAppLocalFiles(String localFilePathList, Optional<Map<String, LocalResource>> resourceMap,
```
just wanted to keep one for single file addition, but ya its mergeable into one. Thanks

jhsenjaliya commented on gobblin-utility/src/main/java/org/apache/gobblin/util/ConfigUtils.java:
```diff
@@ -67,12 +70,48 @@
   public static final String STRIP_SUFFIX = ".ROOT_VALUE";
 
+  private static volatile Config instance;
+  public static Config getInstance() {
```
the empty wont have the config from application.conf and i notice we keep loading config many times in code. so this method should set the base for all initial config and provide the singleton instead of loading the config multiple times.

jhsenjaliya commented on gobblin-utility/src/main/java/org/apache/gobblin/util/ConfigUtils.java:
```diff
@@ -569,4 +608,124 @@ public static Config resolveEncrypted(Config config, Optional<String> encConfigP
+   * @param config at path: {@link ConfigUtils#GOBBLIN_SYNC_SYSTEMS_KEY}
+   * @return map of system name to that system's config tree for all the systems that has security enabled
+   */
+  public static Map<String, ConfigObject> getAllSecureSystems(Config config) {
```
I am taking `configUtils` as a `GobblinConfigUtils` which should have config utility functions with gobblin's context. but sure it can be separated out.

jhsenjaliya commented on gobblin-utility/src/main/java/org/apache/gobblin/util/ConfigUtils.java:
```diff
@@ -569,4 +608,124 @@ public static Config resolveEncrypted(Config config, Optional<String> encConfigP
+    for (Map.Entry<String, ConfigValue> sysConfig : allSysConfig.entrySet()) {
+
+      if (sysConfig.getValue().valueType().equals(ConfigValueType.OBJECT)) {
+        ConfigObject allComponentConfigs = (ConfigObject) sysConfig.getValue();
```
this makes the traversing possible, only `ConfigObject` provides the entrySet method which `ConfigValue` does not. I think there can be other ways to obtain `ConfigObject` out of `ConfigValue` as well.

jhsenjaliya committed:
[GOBBLIN-1398] updates as per the review comments

jhsenjaliya commented:
@autumnust , addressed all comments, can u pls take a look?
about the testcase, we dont have much structure in place for it, but I will take a look at `GobblinYarnAppLauncherTest` and see if we can add it. Thanks
